pkg/reconciler/awscognitosource/controller.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright (c) 2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awscognitosource
+
+import (
+	"context"
+
+	"github.com/kelseyhightower/envconfig"
+
+	"k8s.io/client-go/tools/cache"
+
+	k8sclient "knative.dev/pkg/client/injection/kube/client"
+	deploymentinformerv1alpha1 "knative.dev/pkg/client/injection/kube/informers/apps/v1/deployment"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+	"knative.dev/pkg/resolver"
+
+	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	"github.com/triggermesh/aws-event-sources/pkg/client/generated/injection/client"
+	informerv1alpha1 "github.com/triggermesh/aws-event-sources/pkg/client/generated/injection/informers/sources/v1alpha1/awscognitosource"
+	reconcilerv1alpha1 "github.com/triggermesh/aws-event-sources/pkg/client/generated/injection/reconciler/sources/v1alpha1/awscognitosource"
+)
+
+// NewController creates a Reconciler for the event source and returns the result of NewImpl.
+func NewController(
+	ctx context.Context,
+	cmw configmap.Watcher,
+) *controller.Impl {
+
+	// Calling envconfig.Process() with a prefix appends that prefix
+	// (uppercased) to the Go field name, e.g. MYSOURCE_IMAGE.
+	adapterCfg := &adapterConfig{}
+	envconfig.MustProcess(adapterName, adapterCfg)
+
+	logger := logging.FromContext(ctx)
+
+	sourceInformer := informerv1alpha1.Get(ctx)
+	deploymentInformer := deploymentinformerv1alpha1.Get(ctx)
+
+	r := &Reconciler{
+		logger:           logger,
+		adapterCfg:       adapterCfg,
+		sourceClient:     client.Get(ctx).SourcesV1alpha1().AWSCognitoSources,
+		deploymentClient: k8sclient.Get(ctx).AppsV1().Deployments,
+		deploymentLister: deploymentInformer.Lister().Deployments,
+	}
+	impl := reconcilerv1alpha1.NewImpl(ctx, r)
+
+	logger.Info("Setting up event handlers.")
+
+	sourceInformer.Informer().AddEventHandler(controller.HandleAll(impl.Enqueue))
+
+	deploymentInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.FilterGroupVersionKind((&v1alpha1.AWSCognitoSource{}).GetGroupVersionKind()),
+		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
+	})
+
+	r.sinkResolver = resolver.NewURIResolver(ctx, impl.EnqueueKey)
+
+	cmw.Watch(logging.ConfigMapName(), r.updateAdapterLoggingConfig)
```

Comment:
I may be wrong, but logging config is handled on the shared controller level:
https://github.com/knative/eventing-contrib/issues/1067

same may be true for metrics config

Reply:
That'd be great! But then where do you get the config from? In your code you set both to `""`, which doesn't seem right.

Reply:
I found where logging config comes from if adapter env doesn't have it:
https://github.com/knative/pkg/blob/master/logging/config.go#L160

but I'm wondering what was the need to set `required:"true"` parameter for `K_LOGGING_CONFIG` env variable and have default fallback in the same time

Reply:
Nice catch. This will always use the default settings though, which means there is no possibility to enable debug logging for example.With the cmw, the changes are propagated down to adapters which is really neat.

Maybe we can find a way to have a _single_ watcher instead of one per reconciler, and propagate the logging config in the main context.

Reply:
I'm still trying to understand why I was asked to remove cmw from gitlabsource; I expected to see that watcher was moved to the pkg/adapter package, but it doesn't seem to be true. Maybe adapter's pod (kservice) lifecycle doesn't assume watching for logging updates... :confused: Anyways

> Maybe we can find a way to have a single watcher instead of one per reconciler, and propagate the logging config in the main context.

yeah, that's what I was thinking too